### PR TITLE
handle data-entries limit

### PR DESCRIPTION
**Pull request by domnikov-timofei**

### CODE REVIEW

File: src/_core/chunk.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+export function chunk<T>(array: T[], count: number) {
```
Comment (smelnikov) [inline src/_core/chunk.ts:1]: Is there any reason why we can't use `slice` as in this case?
https://github.com/Keeper-Wallet/Keeper-Wallet-Extension/blob/KEEP-1063/handle-data-entries-limit/src/controllers/assetInfo.ts#L352:L367

File: src/nfts/utils.ts [hunk 2]
```diff
@@ -4,12 +4,11 @@ export function capitalize(str: string | undefined) {
...
-export function reduceDataEntries(entries: DataTransactionEntry[]) {
-  return entries.reduce<Record<string, DataTransactionEntry['value']>>(
-    (data, item) => {
-      data[item.key] = item.value;
-      return data;
-    },
-    {}
-  );
+export function reduceDataEntries<
+  T extends { key: DataTransactionEntry['key']; value: unknown }
```
Comment (smelnikov) [inline src/nfts/utils.ts:8]: Actually, the `entries` type is `DataTransactionEntry[]`.
Why are we omitting `type`?

File: src/_core/chunk.ts (hunk 1 above)
Comment (domnikov-timofei) [reply src/_core/chunk.ts:1]: maybe something like this?

```javascript
function* chunk(array, size) {
  for (const i = 0; i < array.length; i += size) {
    yield array.slice(i, i + size);
  }
}
```

seems to be easier to read)

File: src/nfts/utils.ts (hunk 2 above)
Comment (domnikov-timofei) [reply src/nfts/utils.ts:8]: `DataTransactionEntry` is a union type that combines `DataTransactionEntryInteger`, `DataTransactionEntryString` and so on. if we pass data of type such as `DataTransactionEntryInteger` to this function, then at the output we will get a type inferred from `DataTransactionEntry`, but we expected it to be processed as `DataTransactionEntryInteger`

File: src/_core/fetchInBatches.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import invariant from 'tiny-invariant';
+
+import { chunk } from './chunk';
+
+export function fetchInBatches<T>(
+  allItems: unknown[],
```
Comment (faergeek) [inline src/_core/fetchInBatches.ts:6]: Why not use a generic for this as well instead of `unknown`?

File: src/_core/fetchInBatches.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  chunkSize: number,
+  fetchFn: (items: unknown[]) => ReturnType<typeof fetch>
+): Promise<T[]> {
+  const chunks = chunk(allItems, chunkSize);
+
+  // eslint-disable-next-line no-async-promise-executor
+  return new Promise<T[]>(async (resolve, reject) => {
```
Comment (faergeek) [inline src/_core/fetchInBatches.ts:13]: Why do you need a promise constructor here at all? Why not just turn this whole `fetchInBatches` function into an async function instead?

This way you won't need to wrap loop contents in a try/catch and `reject` and `break` in `catch`. Let it just throw instead :-)

File: src/_core/fetchInBatches.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import invariant from 'tiny-invariant';
+
+import { chunk } from './chunk';
+
+export function fetchInBatches<T>(
+  allItems: unknown[],
+  chunkSize: number,
+  fetchFn: (items: unknown[]) => ReturnType<typeof fetch>
+): Promise<T[]> {
```
Comment (faergeek) [inline src/_core/fetchInBatches.ts:9]: `fetchFn` should return `Promise<T[]>` instead. This function doesn't need to know anything about `fetch`

File: src/nfts/utils.ts (hunk 2 above)
Comment (faergeek) [reply src/nfts/utils.ts:8]: Just `T extends DataTransactionEntry` would work as well

File: src/nfts/vendors/signArt.ts [hunk 6]
```diff
@@ -124,7 +126,8 @@ export class SignArtNftVendor implements NftVendor<SignArtNftInfo> {
           const artDisplayCid = artworksEntries[entriesPerAsset * index + 2];
           const userName = userNameEntries[index];
           const [, artworkId, creator] =
-            artName.key.match(/art_name_(\w+)_(\w+)/i);
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            artName.key.match(/art_name_(\w+)_(\w+)/i)!;
```
Comment (faergeek) [inline src/nfts/vendors/signArt.ts:130]: Better use `invariant` instead of type assertion

**Review by faergeek — CHANGES_REQUESTED** [review]
I think it would be better to implement a function like `fetchDataEntries`, which will:

1. Use `fetchInBatches` internally. Node api limits for data entries doesn't change between different requests anyway, so we don't need to pass it every time.
2. Do what `reduceDataEntries` does, meaning it will always return a record back instead of an array.

What do you think?

**Review by faergeek — APPROVED** [review]

**Review by smelnikov — APPROVED** [review]

**Pull request merged**